src/Raven.Server/Documents/Replication/Outgoing/AbstractOutgoingReplicationHandler.cs

```diff
@@ -0,0 +1,804 @@
+                if (_cts.IsCancellationRequested == false)
+                {
+                    OnFailed(e);
+                }
+            }
+
+            void HandleIOException(IOException e)
+            {
+                if (Logger.IsInfoEnabled)
+                {
+                    if (e.InnerException is SocketException)
+                        Logger.Info($"SocketException was thrown from the connection to remote node ({FromToString}). " +
+                                    $"This might mean that the remote node is done or there is a network issue.", e);
+                    else
+                        Logger.Info($"IOException was thrown from the connection to remote node ({FromToString}).", e);
+                }
+                OnFailed(e);
+            }
+
+            void HandleLegacyReplicationViolationException(LegacyReplicationViolationException e)
+            {
+                if (Logger.IsInfoEnabled)
+                    Logger.Info($"LegacyReplicationViolationException occurred on replication thread ({FromToString}). " +
+                                "Replication is stopped and will not continue until the violation is resolved. ", e);
+                OnFailed(e);
+            }
+
+            void HandleException(Exception e)
+            {
+                if (Logger.IsInfoEnabled)
+                    Logger.Info($"Unexpected exception occurred on replication thread ({FromToString}). " +
+                                $"Replication stopped (will be retried later).", e);
+                OnFailed(e);
+            }
+        }
+
+        protected virtual DynamicJsonValue GetSendPreliminaryDataRequest()
+        {
+            return new DynamicJsonValue
+            {
+                ["Type"] = nameof(ReplicationInitialRequest)
+            };
+        }
+        private void SendPreliminaryData()
+        {
+            var request = GetSendPreliminaryDataRequest();
+
+            using (_contextPool.AllocateOperationContext(out JsonOperationContext context))
+            using (var writer = new BlittableJsonTextWriter(context, _stream))
+            {
+                context.Write(writer, request);
+                writer.Flush();
+            }
+        }
+
+        protected void AddReplicationPulse(ReplicationPulseDirection direction, string exceptionMessage = null)
+        {
+            HandleReplicationPulse?.Invoke(new LiveReplicationPulsesCollector.ReplicationPulse
+            {
+                OccurredAt = SystemTime.UtcNow,
+                Direction = direction,
+                To = Destination,
+                IsExternal = GetType() != typeof(OutgoingInternalReplicationHandler),
+                ExceptionMessage = exceptionMessage
+            });
+        }
+
+        protected abstract void InitiatePullReplicationAsSink(TcpConnectionHeaderMessage.SupportedFeatures socketResultSupportedFeatures, X509Certificate2 certificate);
```

Comment:
can we move those abstract methods to the same place, now you have them in two places in this file. It is more readable if they are one after another.

Reply:
done